splice_machine/src/main/java/com/splicemachine/derby/impl/sql/compile/BroadcastJoinStrategy.java

```diff
@@ -143,7 +143,7 @@ public void estimateCost(Optimizable innerTable,
             double joinCost = broadcastJoinStrategyLocalCost(innerCost, outerCost, totalJoinedRows);
             innerCost.setLocalCost(joinCost);
             innerCost.setLocalCostPerPartition(joinCost);
-            innerCost.setRemoteCost(SelectivityUtil.getTotalRemoteCost(innerCost, outerCost, totalOutputRows));
+            innerCost.setRemoteCost(SelectivityUtil.getTotalPerPartitionRemoteCost(innerCost, outerCost, totalOutputRows));
```

Comment:
I think we shouldn't set remote cost here with per partition cost. For broadcast join and nested loop join, we do need to transfer the qualified rows from the whole right tables. So the transfer cost as part of the join computation in broadcastJoinStrategyLocalCost() is correct to use innerCost.remoteCost() which represents the whole table but not a partition. Similarly, if we look at the mergeSortJonStrategyLocalCost(), we amortize the remote cost by the number of partitions during the computation of shuffle cost.

Reply:
@yxia92 and I discussed the issues here.  localCost is getting divided by partitionCount after each join, which makes the joinCost stay very small.  I'll be updating the join costing to always use localCostPerPartition() to avoid this, and will also be adding remoteCostPerPartition().

Reply:
@yxia92 I pushed an update to save remoteCostPerPartition, and also make sure remoteCostPerPartition and localCostPerPartition are calculated in all code paths.  Please take a look at your convenience.  I am running TPCH 10T with the changes now.